### PR TITLE
EZP-XXXXX: Lookup Limitations

### PR DESCRIPTION
**Pull request by mikadamczyk**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-XXXXX](https://jira.ez.no/browse/EZP-XXXXX)
| **Bug/Improvement**| yes/no
| **New feature**    | yes/no
| **Target version** | `6.x`/`7.x` for bug fixes or improvements _(on existing features)_, `master` for features
| **BC breaks**      | yes/no
| **Tests pass**     | yes/no
| **Doc needed**     | yes/no

**TODO**:
- [ ] Implement feature / fix a bug.
- [ ] Implement tests.
- [ ] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [ ] Ask for Code Review.

### CODE REVIEW

File: eZ/Publish/Core/Repository/Permission/PermissionResolver.php [hunk 1]
```diff
@@ -241,6 +242,113 @@ public function canUser($module, $function, ValueObject $object, array $targets
...
+     * @param \eZ\Publish\API\Repository\Values\User\UserReference $currentUserReference
+     * @param \eZ\Publish\API\Repository\Values\ValueObject $object
+     * @param array $targets
+     *
+     * @return bool
+     *
+     * @throws \eZ\Publish\API\Repository\Exceptions\BadStateException
+     * @throws \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException
+     */
+    private function deniedByRoleLimitation(?Limitation $limitation, APIUserReference $currentUserReference, ValueObject $object, ?array $targets): bool
```
Comment (adamwojs) [inline eZ/Publish/Core/Repository/Permission/PermissionResolver.php]:
```suggestion
    private function isDeniedByRoleLimitation(?Limitation $limitation, APIUserReference $currentUserReference, ValueObject $object, ?array $targets): bool
```

File: eZ/Publish/Core/Repository/Permission/PermissionResolver.php [hunk 2]
```diff
@@ -241,6 +242,113 @@ public function canUser($module, $function, ValueObject $object, array $targets
...
+     * @param \eZ\Publish\API\Repository\Values\User\UserReference $currentUserReference
+     * @param \eZ\Publish\API\Repository\Values\ValueObject $object
+     * @param array $targets
+     *
+     * @return bool
+     *
+     * @throws \eZ\Publish\API\Repository\Exceptions\BadStateException
+     * @throws \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException
+     */
+    private function grantedByLimitation(Limitation $limitation, APIUserReference $currentUserReference, ValueObject $object, ?array $targets): bool
```
Comment (adamwojs) [inline eZ/Publish/Core/Repository/Permission/PermissionResolver.php]:
```suggestion
    private function isGrantedByLimitation(Limitation $limitation, APIUserReference $currentUserReference, ValueObject $object, ?array $targets): bool
```

File: eZ/Publish/API/Repository/Values/User/LookupPolicyLimitations.php [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace eZ\Publish\API\Repository\Values\User;
+
+use eZ\Publish\API\Repository\Values\ValueObject;
+
+/**
+ * This class represents a LookupLimitation for module and function in the context of current User.
```
Comment (adamwojs) [inline eZ/Publish/API/Repository/Values/User/LookupPolicyLimitations.php]:
```suggestion
 * This class represents a result of lookup limitation for module and function in the context of current User.
```

File: eZ/Publish/API/Repository/Values/User/LookupPolicyLimitations.php [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+{
+    /** @var \eZ\Publish\API\Repository\Values\User\Policy */
+    protected $policy;
+
+    /** @var \eZ\Publish\API\Repository\Values\User\Limitation[] */
+    protected $limitations;
+
+    /**
+     * @param \eZ\Publish\API\Repository\Values\User\Policy $policy
+     * @param array $limitations
```
Comment (adamwojs) [inline eZ/Publish/API/Repository/Values/User/LookupPolicyLimitations.php]:
```suggestion
     * @param \eZ\Publish\API\Repository\Values\User\Limitation[] $limitations
```

File: eZ/Publish/Core/Repository/Permission/PermissionResolver.php [hunk 5]
```diff
@@ -241,6 +242,113 @@ public function canUser($module, $function, ValueObject $object, array $targets
...
+        }
+
+        return new LookupLimitationResult(!empty($passedLimitations), $passedLimitations);
+    }
+
+    /**
+     * @param \eZ\Publish\API\Repository\Values\User\Limitation $limitation
+     * @param \eZ\Publish\API\Repository\Values\User\UserReference $currentUserReference
+     * @param \eZ\Publish\API\Repository\Values\ValueObject $object
+     * @param array $targets
```
Comment (adamwojs) [inline eZ/Publish/Core/Repository/Permission/PermissionResolver.php]:
```suggestion
     * @param array|null $targets
```

File: eZ/Publish/Core/Repository/Permission/PermissionResolver.php [hunk 6]
```diff
@@ -241,6 +242,113 @@ public function canUser($module, $function, ValueObject $object, array $targets
...
+        $accessVote = $type->evaluate($limitation, $currentUserReference, $object, $targets);
+
+        return $accessVote === LimitationType::ACCESS_GRANTED;
+    }
+
+    /**
+     * @param \eZ\Publish\API\Repository\Values\User\Limitation|null $limitation
+     * @param \eZ\Publish\API\Repository\Values\User\UserReference $currentUserReference
+     * @param \eZ\Publish\API\Repository\Values\ValueObject $object
+     * @param array $targets
```
Comment (adamwojs) [inline eZ/Publish/Core/Repository/Permission/PermissionResolver.php]:
```suggestion
     * @param array|null $targets
```

File: eZ/Publish/API/Repository/Values/User/LookupLimitationResult.php [hunk 7]
```diff
@@ -0,0 +1,37 @@
+<?php
+
+/**
+ * File containing the eZ\Publish\API\Repository\Values\User\LookupLimitation class.
```
Comment (adamwojs) [inline eZ/Publish/API/Repository/Values/User/LookupLimitationResult.php]:
```suggestion
```

File: eZ/Publish/API/Repository/Values/User/LookupPolicyLimitations.php [hunk 8]
```diff
@@ -0,0 +1,37 @@
+<?php
+
+/**
+ * File containing the eZ\Publish\API\Repository\Values\User\LookupPolicyLimitations class.
```
Comment (adamwojs) [inline eZ/Publish/API/Repository/Values/User/LookupPolicyLimitations.php]:
```suggestion
```

File: eZ/Publish/Core/Repository/Permission/PermissionResolver.php [hunk 9]
```diff
@@ -241,6 +242,113 @@ public function canUser($module, $function, ValueObject $object, array $targets
...
+
+        $currentUserReference = $this->getCurrentUserReference();
+
+        $passedLimitations = [];
+        foreach ($permissionSets as $permissionSet) {
+            if ($this->deniedByRoleLimitation($permissionSet['limitation'], $currentUserReference, $object, $targets)) {
+                continue;
+            }
+
+            /** @var $policy \eZ\Publish\API\Repository\Values\User\Policy */
```
Comment (adamwojs) [inline eZ/Publish/Core/Repository/Permission/PermissionResolver.php]:
```suggestion
            /** @var \eZ\Publish\API\Repository\Values\User\Policy $policy */
```

**Comment by alongosz** [comment]
Moved to #2585 due to time constraints (v2.5.0 final approaching fast).

**Pull request closed without merging**